pyrpl/redpitaya.py

```diff
@@ -292,41 +318,78 @@ def update_fpga(self, filename=None):
+            os.path.join(self.parameters['serverdirname'], 'update_fpga.sh'))
+        # send binfile
+        self.put_file(source, os.path.join(self.parameters['serverdirname'],
+                                           self.parameters['serverbinfilename']))
+        if 0 < len(update_custom) and 'dtbo_filename' in self.parameters and not self.parameters['dtbo_filename'] is None:
+            # send device tree
+            update_custom = '{} dtbo'.format(update_custom)
+            if not os.path.isfile(self.parameters['dtbo_filename']):
+                raise IOError("Wrong device tree filename",
+                  "The fpga device tree was not found at the expected location. Try passing the arguments "
+                  "dtbo_filename=\"c://github//pyrpl//pyrpl//red_pitaya.dtbo\" adapted to your installation "
+                  "directory of pyrpl: current filename: "+self.parameters['dtbo_filename'])
+            self.put_file(self.parameters['dtbo_filename'],
+                          os.path.join(self.parameters['serverdirname'],
+                                       self.parameters['serverdtbofilename']))
+
+        # add appropriate parameters to the update command line
+        update_cmd = '{} {}'.format(update_cmd, update_custom)
+
         # kill all other servers to prevent reading while fpga is flashed
         self.end()
         self.ssh.ask('killall nginx')
         self.ssh.ask('systemctl stop redpitaya_nginx') # for 0.94 and higher
-        self.ssh.ask('cat '
-                 + os.path.join(self.parameters['serverdirname'], self.parameters['serverbinfilename'])
-                 + ' > //dev//xdevcfg')
+        sleep(3) # sleep after stopping service
+        result = self.ssh.ask('cat /root/.version')
+        self.logger.debug('cat /root/.version: {}'.format(result))
+        if result.find('2.') != -1:
+            self.ssh.ask(update_cmd)
+            sleep(1)
+        else:
+            self.ssh.ask('cat '
+                         + os.path.join(self.parameters['serverdirname'], self.parameters['serverbinfilename'])
+                         + ' > //dev//xdevcfg')
+
         sleep(self.parameters['delay'])
-        self.ssh.ask('rm -f '+ os.path.join(self.parameters['serverdirname'], self.parameters['serverbinfilename']))
+        self.logger.debug('About to restart the redpitaya service')
+        self.ssh.ask('rm -f '+ os.path.join(
+            self.parameters['serverdirname'], self.parameters['serverdtbofilename']))
+        self.ssh.ask('rm -f '+ os.path.join(
+            self.parameters['serverdirname'], self.parameters['serverbinfilename']))
+        self.ssh.ask('rm -f '+ os.path.join(
+            self.parameters['serverdirname'], 'update_fpga.sh'))
         self.ssh.ask("nginx -p //opt//www//")
-        self.ssh.ask('systemctl start redpitaya_nginx')  # for 0.94 and higher #needs test
+        self.ssh.ask('systemctl start redpitaya_nginx')  # for 0.94 and higher
```

Comment:
```suggestion
        self.ssh.ask('systemctl stop redpitaya_nginx')  # for 0.94 and higher
```
I'd just try to never have this running together with pyrpl

Reply:
I have found it useful to swap between pyrpl and "normal" operation with having to reboot the RedPitaya that wastes quite a bit of time waiting for the reboot. In my opinion if we stop the nginx server we should provide an API to re-start it or should guarantee to restart the nginx server when pyrpl is stopped.

Note that the file /tmp/loaded_fpga.inf (and update_fpga.txt) can be parsed to detect if the "normal" apps are being used so in the event of failure to read pyrpl registers remedial action can be taken by the pyrpl code (ie exit if "normal" apps is in use with appropriate message to the pyrpl user).  That type of action might be more user friendly!

what do you think?